fairscale/optim/oss.py

```diff
@@ -517,20 +454,75 @@ def add_param_group(self, param_group: dict) -> None:
             # Update the bucketing strategy accordingly
             self._setup_flat_buffers()
 
+    @property
+    def _local_params(self) -> List[torch.Tensor]:
+        """ Iterable which goes through the parameters that this rank owns
+        """
+        if self.__local_params is None:
+            self.__local_params = list(
+                chain(
+                    *[
+                        list(filter(lambda x: x.grad is not None, device_params[self.rank]))
+                        for device_params in self._per_device_params.values()
+                    ]
+                )
+            )
+
+        # Make sure that the iterator is not consumed, only expose a copy
+        return self.__local_params
+
+    @property
+    def _param_to_index(self) -> Dict[int, int]:
+        """ Hash table in between parameter indices in the global optimizer scheme, and the actual params
+        """
+        if len(self.__param_to_index) == 0:
+            self.__param_to_index = {id(p): i for i, p in enumerate(chain(*(g["params"] for g in self.param_groups)))}
+
+        return self.__param_to_index
+
+    @property
+    def _per_device_params(self) -> Dict[torch.device, List[List[Parameter]]]:
+        """Sorted list of all the params, first per device then per rank.
+
+        Within a list params are sorted per number of elements to allow for an easy bucketing.
+        """
+        if len(self.__per_device_params) == 0:
+            # Go through all params, log them per device
+            # The ordering is important here, needs to be the same on all ranks
+            # So that ulterior broadcast calls are matching
+            for param_group in self.param_groups:
+                for param in param_group["params"]:
+                    device = param.device
+                    if self.__per_device_params.get(device) is None:
+                        self.__per_device_params[device] = [[] for _ in range(self.world_size)]
+                    self.__per_device_params[device][self._param_to_rank[param]] += [param]
+
+            # Sort param_lists by size
+            for device in self.__per_device_params.keys():
+                for rank_params in self.__per_device_params[device]:
+                    rank_params.sort(key=lambda x: x.numel())
+
+        return self.__per_device_params
+
+    @property
+    def _param_to_rank(self) -> Dict[torch.Tensor, int]:
+        """param to data parallel rank"""
```

Comment:
perhaps a slightly more verbose string that mentions which "param" would be good.

Reply:
I tried to explain a little more, this applies to all params actually :)